torchrl/envs/libs/jumanji.py

```diff
@@ -0,0 +1,430 @@
+            dtype = numpy_to_torch_dtype_dict[spec.dtype]
+        if dtype in (torch.float, torch.double, torch.half):
+            return NdUnboundedContinuousTensorSpec(
+                shape=spec.shape, dtype=dtype, device=device
+            )
+        else:
+            return NdUnboundedDiscreteTensorSpec(
+                shape=spec.shape, dtype=dtype, device=device
+            )
+    elif isinstance(spec, jumanji.specs.Spec) and hasattr(spec, "__dict__"):
+        new_spec = {}
+        for key, value in spec.__dict__.items():
+            if isinstance(value, jumanji.specs.Spec):
+                if key.endswith("_obs"):
+                    key = key[:-4]
+                if key.endswith("_spec"):
+                    key = key[:-5]
+                new_spec[key] = _jumanji_to_torchrl_spec_transform(
+                    value, dtype, device, categorical_action_encoding
+                )
+        return CompositeSpec(**new_spec)
+    else:
+        raise TypeError(f"Unsupported spec type {type(spec)}")
+
+
+def _data_to_spec_transform(state) -> TensorSpec:
+    if isinstance(state, torch.Tensor):
+        if state.dtype in (torch.float, torch.double, torch.half):
+            return NdUnboundedContinuousTensorSpec(
+                shape=state.shape, dtype=state.dtype, device=state.device
+            )
+        else:
+            return NdUnboundedDiscreteTensorSpec(
+                shape=state.shape, dtype=state.dtype, device=state.device
+            )
+    elif isinstance(state, TensorDict):
+        return CompositeSpec(
+            **{key: _data_to_spec_transform(value) for key, value in state.items()}
+        )
+    else:
+        raise TypeError(f"Unsupported state type {type(state)}")
+
+
+def _jumanji_to_torchrl_data_transform(val, device, batch_size):
+    if isinstance(val, (jax.Array, np.ndarray)):
+        if isinstance(val, jax.Array):
+            val = np.array(val)
+        if val.dtype == np.uint16:
+            val = val.astype(np.int16)
+        if val.dtype == np.uint32:
+            val = val.astype(np.int32)
+        if val.dtype == np.uint64:
+            val = val.astype(np.int64)
+        return torch.tensor(val, device=device)
+    elif isinstance(val, tuple) and hasattr(val, "_fields"):  # named tuples
+        return TensorDict(
+            {
+                k: _jumanji_to_torchrl_data_transform(v, device, batch_size)
+                for k, v in zip(val._fields, val)
+            },
+            device=device,
+            batch_size=batch_size,
+        )
+    elif hasattr(val, "__dict__"):
+        return TensorDict(
+            {
+                k: _jumanji_to_torchrl_data_transform(v, device, batch_size)
+                for k, v in val.__dict__.items()
+            },
+            device=device,
+            batch_size=batch_size,
+        )
+    else:
+        raise TypeError(f"Unsupported data type {type(val)}")
+
+
+def _torchrl_to_jumanji_state_transform(tensordict: TensorDict, env):
```

Comment:
can't we use the reverse function of the one above?

Reply:
right, this will be refactored.